File: src/repotask/connectors/secrets.py
"""Credential lookup for REST connectors.

Credentials never live in the project: they come from the environment first, then
`~/.repo-task/secrets.yaml`, which is outside version control by construction.
"""

from __future__ import annotations

import os
from functools import lru_cache
from pathlib import Path

import yaml

from repotask.errors import RepoTaskError

SECRETS_ENV = "REPOTASK_SECRETS_FILE"


def secrets_path() -> Path:
    override = os.environ.get(SECRETS_ENV)
    return Path(override).expanduser() if override else Path.home() / ".repo-task/secrets.yaml"


@lru_cache(maxsize=1)
def _load() -> dict[str, dict[str, str]]:
    path = secrets_path()
    if not path.is_file():
        return {}
    raw = yaml.safe_load(path.read_text(encoding="utf-8")) or {}
    if not isinstance(raw, dict):
        raise RepoTaskError(f"{path} must contain a YAML mapping of system -> credentials.")
    return {
        str(system): {str(key): str(value) for key, value in values.items()}
        for system, values in raw.items()
        if isinstance(values, dict)
    }


def get(system: str, key: str) -> str:
    """Look up `key` for `system`; environment wins over the secrets file."""
    env_name = f"REPOTASK_{system.upper().replace('-', '_')}_{key.upper()}"
    return os.environ.get(env_name) or _load().get(system, {}).get(key, "")


def require(system: str, key: str) -> str:
    value = get(system, key)
    if not value:
        env_name = f"REPOTASK_{system.upper().replace('-', '_')}_{key.upper()}"
        raise RepoTaskError(
            f"Missing credential '{key}' for '{system}'. Set {env_name} or add it under "
            f"'{system}:' in {secrets_path()}."
        )
    return value


def reset_cache() -> None:
    _load.cache_clear()

File: src/connectors/secrets.rs
//! Credential lookup for REST connectors.
//!
//! Credentials never live in the project: they come from the environment first, then
//! `~/.repo-task/secrets.yaml`, which is outside version control by construction.

use std::collections::BTreeMap;
use std::path::PathBuf;

use anyhow::{bail, Result};

pub const SECRETS_ENV: &str = "REPOTASK_SECRETS_FILE";

pub fn secrets_path() -> PathBuf {
    if let Ok(override_path) = std::env::var(SECRETS_ENV) {
        return PathBuf::from(override_path);
    }
    let home = std::env::var("HOME")
        .or_else(|_| std::env::var("USERPROFILE"))
        .unwrap_or_else(|_| ".".into());
    PathBuf::from(home).join(".repo-task/secrets.yaml")
}

fn load() -> BTreeMap<String, BTreeMap<String, String>> {
    let path = secrets_path();
    let Ok(text) = std::fs::read_to_string(&path) else {
        return BTreeMap::new();
    };
    serde_norway::from_str(&text).unwrap_or_default()
}

fn env_name(system: &str, key: &str) -> String {
    format!(
        "REPOTASK_{}_{}",
        system.to_uppercase().replace('-', "_"),
        key.to_uppercase()
    )
}

/// Look up `key` for `system`; environment wins over the secrets file.
pub fn get(system: &str, key: &str) -> String {
    if let Ok(value) = std::env::var(env_name(system, key)) {
        if !value.is_empty() {
            return value;
        }
    }
    load()
        .get(system)
        .and_then(|values| values.get(key))
        .cloned()
        .unwrap_or_default()
}

pub fn require(system: &str, key: &str) -> Result<String> {
    let value = get(system, key);
    if value.is_empty() {
        bail!(
            "Missing credential '{key}' for '{system}'. Set {} or add it under '{system}:' in {}.",
            env_name(system, key),
            secrets_path().display()
        );
    }
    Ok(value)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn environment_variable_names_are_predictable() {
        assert_eq!(env_name("jira", "token"), "REPOTASK_JIRA_TOKEN");
        assert_eq!(
            env_name("github-enterprise", "token"),
            "REPOTASK_GITHUB_ENTERPRISE_TOKEN"
        );
    }

    #[test]
    fn missing_credentials_name_both_sources() {
        let error = require("jira", "token").unwrap_err().to_string();

        assert!(error.contains("REPOTASK_JIRA_TOKEN"), "{error}");
        assert!(error.contains("secrets.yaml"), "{error}");
    }
}
